src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs

```diff
@@ -780,70 +785,113 @@ private async Task<(IPackageFeed mainFeed, IPackageFeed recommenderFeed)> GetPac
-
-            if (pSearchCallback != null && searchTask != null)
+            try
             {
-                var searchResult = await searchResultTask;
-                pSearchCallback.ReportComplete(searchTask, (uint)searchResult.RawItemsCount);
-            }
+                var loader = new PackageItemLoader(
+                    loadContext, packageFeeds.mainFeed, searchText, IncludePrerelease, packageFeeds.recommenderFeed);
+                var loadingMessage = string.IsNullOrWhiteSpace(searchText)
+                    ? Resx.Resources.Text_Loading
+                    : string.Format(CultureInfo.CurrentCulture, Resx.Resources.Text_Searching, searchText);
 
-            // We only refresh update count, when we don't use cache so check if it's false
-            if (!useCacheForUpdates)
-            {
-                // clear existing caches
-                Model.CachedUpdates = null;
+                // Set a new cancellation token source which will be used to cancel this task in case
+                // new loading task starts or manager ui is closed while loading packages.
+                _loadCts = new CancellationTokenSource();
 
-                if (_topPanel.Filter.Equals(ItemFilter.UpdatesAvailable))
-                {
-                    // it means selected tab is update itself, so just wait for searchAsyncTask to complete
-                    // without making another call to loader to get all packages.
-                    _topPanel.UpdateCountOnUpdatesTab(count: 0);
+                // start SearchAsync task for initial loading of packages
+                var searchResultTask = loader.SearchAsync(continuationToken: null, cancellationToken: _loadCts.Token);
+                // this will wait for searchResultTask to complete instead of creating a new task
+                await _packageList.LoadItemsAsync(loader, loadingMessage, _uiLogger, searchResultTask, _loadCts.Token);
 
+                if (pSearchCallback != null && searchTask != null)
+                {
                     var searchResult = await searchResultTask;
-                    Model.CachedUpdates = new PackageSearchMetadataCache
-                    {
-                        Packages = searchResult.Items,
-                        IncludePrerelease = IncludePrerelease
-                    };
-
-                    _topPanel.UpdateCountOnUpdatesTab(Model.CachedUpdates.Packages.Count);
+                    pSearchCallback.ReportComplete(searchTask, (uint)searchResult.RawItemsCount);
                 }
-                else
+
+                // When not using Cache, refresh all Counts.
+                if (!useCachedPackages)
                 {
                     RefreshInstalledAndUpdatesTabs();
                 }
+
+                FlagTabAsLoaded(filterToRender);
+
+                //Loading Installed tab should also consider the Updates tab as loaded to indicate UI filtering is ready.
```

Comment:
perhaps more data and less tabs?

//Loading Installed data should also consider the Updates data as loaded to indicated UI filtering is ready.

Reply:
Are you saying to do this or just "maybe"?

Data to me is more on the Package Source side. Really the word TAB means the UI layer is loaded and ready to go. It happens to mean data is also loaded, but could mean other things unrelated to the data/cache.

Reply:
this all ties into the naming of many variables around Tabs being loaded...but really i think you are more worried about whether the data is loaded.


---
In reply to: [450530538](https://github.com/NuGet/NuGet.Client/pull/3444#discussion_r450530538) [](ancestors = 450530538)

Reply:
Sounds that you're strongly suggesting Data, so I'll rename that part.

Reply:
As a compromise, I named everything as TabData.... like _installedTabDataIsLoaded. Sound good?